### PR TITLE
feat: add config and messaging base of contracts

### PR DESCRIPTION
**Pull request by glihm**
- [X] issue # (WIP of blobstream L3 contract)
- [X] follows contribution [guide](../CONTRIBUTING.md)
- [X] code change includes tests
- [X] breaking change

This PR aims at providing a base impl for appchain core contract on Starknet.

On-going tasks:
- [X] simple base contract (`appchain.cairo`)
- [X] configuration component
- [X] port messaging
- [X] complete messaging with cancelling
- [X] add first state update with tests without facts registry for messages

We will open issues once this is done for contributions and missing features.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,3 +41,36 @@ To test the project, run:
...
+```rust
+// Prefer that:
+let addr = starknet::contract_address_const::<0>();
+
+// Instead of:
+use starknet::contract_address_const;
+let addr = contract_address_const::<0>();
+```
+
+- Order import as much as you can (and avoid relative import, if not inside tests).
```
Comment (drspacemn) [inline README.md:60]: Can enforce to some degree in `Scarb.toml`

```
[tool.fmt]
sort-module-level-items = true
```

Comment (glihm) [reply README.md:60]: I've tried it, it seems to only sort by alphabetic order and does not respect the new lines as separator all the time.

Perhaps for now let's go with that, and group the imports. Will remove this guideline, sounds good to you?

**Comment by glihm** [comment]
@drspacemn I think we're good for a first PR and review before adding more stuff.

Next PR will be for rust bindings.
And after that we can check the state part. :+1:

File: src/interface.cairo [hunk 2]
```diff
@@ -0,0 +1,53 @@
+//! SPDX-License-Identifier: MIT
```
Comment (drspacemn) [inline src/interface.cairo:1]: This is the same interface as the `IConfig`?

File: src/messaging/output_process.cairo [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    /// Nonce.
+    nonce: felt252,
+    /// Function selector (with #[l1 handler] attribute).
+    selector: felt252,
+    /// Payload size.
+    payload: Span<felt252>,
+}
+
+const MESSAGE_TO_STARKNET_HEADER_SIZE: usize = 3;
+const MESSAGE_TO_APPCHAIN_HEADER_SIZE: usize = 5;
```
Comment (drspacemn) [inline src/messaging/output_process.cairo:41]: let's move to the snos_output.cairo?

**Review by drspacemn — COMMENTED** [review]
For `blobstream_starknet` we include unit `tests` in the module and then the top level tests are for the contract logic itself. An example here would be:

`tests/test_appchain.cairo`
`src/config/tests/test_config.cairo`
`src/messaging/tests/test_messaging.cairo`

File: src/appchain.cairo [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    use openzeppelin::access::ownable::{OwnableComponent as ownable_cpt, interface::IOwnable};
+    use piltover::config::{config_cpt, config_cpt::InternalTrait as ConfigInternal, IConfig};
+    use piltover::messaging::{
+        messaging_cpt, messaging_cpt::InternalTrait as MessagingInternal, IMessaging,
+        output_process, output_process::{MessageToStarknet, MessageToAppchain},
+    };
+    use starknet::ContractAddress;
+    use super::errors;
+
+    const SNOS_OUTPUT_HEADER_SIZE: usize = 5;
```
[the remark this answers is not in the document]
Comment (drspacemn) [reply src/appchain.cairo:23]: example `snos_output::HEADER_SIZE`

File: src/appchain.cairo [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    /// Initializes the contract.
+    ///
+    /// # Arguments
+    ///
+    /// * `address` - The contract address of the owner.
+    #[constructor]
+    fn constructor(ref self: ContractState, owner: ContractAddress) {
+        self.ownable.transfer_ownership(owner);
```
[the remark this answers is not in the document]
Comment (glihm) [reply src/appchain.cairo:60]: Yes totally, good catch.

File: src/appchain.cairo [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+        let cancellation_delay_secs = 432000;
```
[the remark this answers is not in the document]
Comment (glihm) [reply src/appchain.cairo:62]: It as something a bit undefined I left here. I'll pass it as module constant yes.

File: src/appchain.cairo (hunk 4 above)
[the remark this answers is not in the document]
Comment (glihm) [reply src/appchain.cairo:23]: Yes, I will update. My header was to update SNOS to use only Cairo serialization. Like this there's no offset or other things to manage.
But in the meantime, I'll update as you mentioned to be closer to solidity's code.

File: src/config/component.cairo [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        interface::IOwnable,
+    };
+    use piltover::config::interface::IConfig;
+    use starknet::ContractAddress;
+    use super::errors;
+
+    #[storage]
+    struct Storage {
+        /// Appchain operator that is allowed to update the state.
+        operator: ContractAddress,
```
[the remark this answers is not in the document]
Comment (glihm) [reply src/config/component.cairo:27]: Was a reduction to simplify it. However to allow better scaling on proving, I'll update to mapping of operators.

File: src/config/component.cairo [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        /// Program info (StarknetOS), with program hash and config hash.
+        program_info: (felt252, felt252),
+        /// Facts registry contract address.
+        facts_registry: ContractAddress,
```
[the remark this answers is not in the document]
Comment (glihm) [reply src/config/component.cairo:31]: As on starknet the facts registry and the verifier will be two distinct contracts, the facts registry contract is already in charge to leverage all the work with the verifier.

From the core contract perspective, we should only require to interact with the fact registry which expose the `is_valid` method: https://github.com/HerodotusDev/cairo-verifier/blob/a30cec152579036bb1abe70b734617da687c7acd/fact_registry/src/lib.cairo#L21.

File: src/interface.cairo (hunk 2 above)
Comment (glihm) [reply src/interface.cairo:1]: This is a file I forgot to remove. x) Sorry for that, good catch.

File: src/config/component.cairo (hunk 8 above)
[the remark this answers is not in the document]
Comment (drspacemn) [reply src/config/component.cairo:31]: Make a lot of sense

File: src/config/component.cairo (hunk 7 above)
[the remark this answers is not in the document]
Comment (drspacemn) [reply src/config/component.cairo:27]: Up to you. I'm fine leaving it as is for simplicity in the short term. Maybe just an issue + TODO to track the work

File: src/appchain.cairo (hunk 5 above)
[the remark this answers is not in the document]
Comment (glihm) [reply src/appchain.cairo]: Fixed in 2e93d1e179c04df1571f7ddce1ca637d3b2c84c9.

File: src/appchain.cairo (hunk 6 above)
[the remark this answers is not in the document]
Comment (glihm) [reply src/appchain.cairo]: Fixed in 9f2cb081bcc7398d78bdd645e858ff115fea3c96.

File: tests/constants.cairo [hunk 9]
```diff
@@ -0,0 +1,31 @@
+//! Test constants.
```
[the remark this answers is not in the document]
Comment (glihm) [reply tests/constants.cairo]: Updated to use OZ constants.

File: src/appchain.cairo (hunk 4 above)
[the remark this answers is not in the document]
Comment (glihm) [reply src/appchain.cairo]: Fixed in fac2c2fc30a3c0799bedfe369a0b7135964ac13d.

File: src/config/component.cairo (hunk 7 above)
[the remark this answers is not in the document]
Comment (glihm) [reply src/config/component.cairo:27]: It's a nice first issue, will open an issue for that to open contributions.

File: Scarb.toml [hunk 10]
```diff
@@ -5,7 +5,12 @@ edition = "2023_10"
 
 [dependencies]
 snforge_std = { git = "https://github.com/foundry-rs/starknet-foundry", tag = "v0.16.0" }
-starknet = "2.5.0"
+starknet = "=2.4.4"
+openzeppelin = { git = "https://github.com/OpenZeppelin/cairo-contracts.git", tag = "v0.8.1" }
```
[the remark this answers is not in the document]
Comment (glihm) [reply Scarb.toml:10]: Fixed in ff896555e1cf058ac29805cc38612b2a5a05eedb.

File: src/interface.cairo (hunk 2 above)
Comment (glihm) [reply src/interface.cairo:1]: Fixed in b7d57341dd35c7ceb18e6216471fe8b5ddf98166.

File: src/messaging/output_process.cairo (hunk 3 above)
Comment (glihm) [reply src/messaging/output_process.cairo]: Fixed in Fixed in fac2c2fc30a3c0799bedfe369a0b7135964ac13d.

**Comment by glihm** [comment]
@drspacemn ready for the second review.
Bumped to 2.5.3, even if OZ has some warnings pending.
Add some testing for appchain `update_state` function.

File: src/appchain.cairo [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    #[abi(embed_v0)]
+    impl Appchain of IAppchain<ContractState> {
+        fn update_state(ref self: ContractState, program_output: Span<felt252>) {
+            self.config.assert_only_owner_or_operator();
+
+            // TODO: reentrancy guard.
+            // TODO: facts verification.
+            // TODO: update the current state (component needed).
```
Comment (drspacemn) [inline src/appchain.cairo:79]: Do you mind adding issues for these?

**Pull request merged**